### PR TITLE
Enable additional features from the existing randomizer

### PR DESCRIPTION
**Pull request by dclaganiere**
- Enabled the following features from the existing randomizer:
  - Plurkwood Reachable
  - Music Shuffle
  - Background Shuffle
  - Gift Item Shuffle
  - Map Transition Shuffle
  - Constraint Changes
- Created basic game and setup pages
  - This was causing unit tests to fail
- Various cleanup and bug fixes

### CODE REVIEW

File: worlds/rabi_ribi/__init__.py [hunk 1]
```diff
@@ -29,6 +32,26 @@ def launch_client():
...
+class RabiRibiWeb(WebWorld):
+    """Web integration for Rabi-Ribi"""
+    tutorials = [Tutorial(
+        "Multiworld Setup Guide",
+        "A guide to setting up Rabi-Ribi integration for Archipelago multiworld games.",
+        "English",
+        "setup_en.md",
+        "setup/en",
+        ["Phie", "PsyMarth"]
+    )]
```
Comment (tdkollins) [inline worlds/rabi_ribi/__init__.py:53]: I'm a fan of leaving every module to do one thing. I would encourage the settings and the web to be in its own dedicated module. I think this is also the convention for most APWorlds within the repo.

File: worlds/rabi_ribi/__init__.py [hunk 2]
```diff
@@ -76,27 +98,63 @@ def create_regions(self) -> None:
         Define regions and locations.
         This also defines access rules for the regions and locations.
         """
+        self.topology_present = bool(self.options.shuffle_map_transitions.value)
+
         region_def = RegionDef(self.multiworld, self.player, self.options)
+
+        # Universal tracker support, can be ignored for standard gen
+        if hasattr(self.multiworld, "re_gen_passthrough") and "Rabi-Ribi" in self.multiworld.re_gen_passthrough: # type: ignore
```
[the remark this answers is not in the document]
Comment (tdkollins) [reply worlds/rabi_ribi/__init__.py:106]: Also suggest possibly pulling this out into its own helper `configure_universal_tracker` to keep create_event as a wrapper function. Keeping these methods simple, short, and descriptive can help a newcomer better parse the code.

File: worlds/rabi_ribi/__init__.py [hunk 3]
```diff
@@ -76,27 +98,63 @@ def create_regions(self) -> None:
...
+        self.picked_templates = region_def.allocation.picked_templates
+        self.map_transition_shuffle_order = region_def.map_transition_shuffle_order
+
+        self.map_transition_shuffle_spoiler: List[str] = []
+        for (idx, x) in enumerate(self.map_transition_shuffle_order):
+            left = region_def.randomizer_data.walking_left_transitions[x]
+            right = region_def.randomizer_data.walking_right_transitions[idx]
+            left_name = convert_existing_rando_name_to_ap_name(left.origin_location)
+            right_name = convert_existing_rando_name_to_ap_name(right.origin_location)
+            self.map_transition_shuffle_spoiler.append(f'{left_name} -> {right_name}')
```
Comment (tdkollins) [inline worlds/rabi_ribi/__init__.py:129]: Similar comment to the above. Would suggest putting this logic in a descriptive region_def method, and keeping this `create_region` method as just a simple wrapper for more complicated region logic.

File: worlds/rabi_ribi/__init__.py [hunk 4]
```diff
@@ -145,3 +219,9 @@ def _handle_encourage_eggs_in_late_spheres(multiworld):
     @classmethod
     def stage_post_fill(cls, multiworld) -> None:
         cls._handle_encourage_eggs_in_late_spheres(multiworld)
+
+    # For the universal tracker, doesn't get called in standard gen
+    @staticmethod
+    def interpret_slot_data(slot_data: Dict[str, Any]) -> Dict[str, Any]:
+        # Returning slot_data so it regens, giving it back in multiworld.re_gen_passthrough
+        return slot_data
```
Comment (tdkollins) [inline worlds/rabi_ribi/__init__.py:227]: Similar confusion about why this is needed. Can we describe the motivation behind this method in a docstring?

File: worlds/rabi_ribi/docs/en_Rabi-Ribi.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+## What does randomization do to this game?
+
+Items which the player would normally acquire throughout the game have been moved around. Logic remains, so the game is
+always able to be completed, but because of the item shuffle, the player may need to access certain areas before they
+would be expected to in the vanilla game.
+
+## What is the goal of Rabi-Ribi in Archipelago
+
+The goal of Rabi-Ribi is to find 5 Easter Egg items shuffled throughout Rabi Rabi Island. Collecting the fifth Easter Egg
+will complete your seed.
```
Comment (tdkollins) [inline worlds/rabi_ribi/docs/en_Rabi-Ribi.md:17]: Suggest making a quick note about how the game can normally be beaten with 0% items, so beating the final boss isn't a very interesting goal (although 5 eggs + boss might be)

File: worlds/rabi_ribi/entrance_shuffle.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            self.error_message = 'Not all locations are reachable.'
+            return False
+
+        return True
+    
+    def verify_reachable_items(self, starting_variables, backward_exitable):
+        # Mark all upgrades as obtained already
+        for item in self.data.must_be_reachable:
+            starting_variables[item] = True
+        return super().verify_reachable_items(starting_variables, backward_exitable)
```
Comment (tdkollins) [inline worlds/rabi_ribi/entrance_shuffle.py:118]: Skimmed this file, but looks fine to me. Did not verify the actual logic though.

**Review by tdkollins — APPROVED** [review]
Left some suggestions. Also suggest keeping these PRs to one logic change per PR in order to keep it simpler to review 😅

File: worlds/rabi_ribi/__init__.py (hunk 1 above)
Comment (tdkollins) [reply worlds/rabi_ribi/__init__.py:53]: interesting... ig I was mistaken haha. ack 👍

File: worlds/rabi_ribi/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dclaganiere) [reply worlds/rabi_ribi/__init__.py:106]: Universal Tracker isn't officially checked in, but it's still widely used to the point that a few games like Undertale and TUNIC has support for it in the main repo (both of which I copied for this implementation). 

I don't want `RegionDef` to be too knowledgeable of Universal Tracker, but my next commit should move more of the responsibility over to `RegionDef` for interpreting the passthrough data, alongside better comments.

Here's a link to the documentation for `re_gen_passthrough` in Universal Tracker (also will be included in comments): https://github.com/FarisTheAncient/Archipelago/blob/tracker/worlds/tracker/docs/re-gen-passthrough.md

File: worlds/rabi_ribi/__init__.py (hunk 3 above)
Comment (dclaganiere) [reply worlds/rabi_ribi/__init__.py:129]: Makes sense

File: worlds/rabi_ribi/__init__.py (hunk 4 above)
Comment (dclaganiere) [reply worlds/rabi_ribi/__init__.py:227]: Updated the documentation here + added a comment with a link to the Universal Tracker documentation.

File: worlds/rabi_ribi/docs/en_Rabi-Ribi.md (hunk 5 above)
Comment (dclaganiere) [reply worlds/rabi_ribi/docs/en_Rabi-Ribi.md:17]: Agreed. I think people who have played the game would be aware that the game is beatable in 0%, but there's def people that do their first play of a game in a randomizer.

**Pull request merged**